File: leetcode/BestTimeToBuyAndSellStockIV.java
public class BestTimeToBuyAndSellStockIV {
  public int maxProfit(int k, int[] prices) {
    int m = prices.length;
    int n = Math.min(k * 2, m);
    if (m == 0) {
      return 0;
    }
    if (n == m) {
      return maxProfit2(prices);
    }
    System.out.printf("m = %d, n = %d\n", m, n);
    // dp[i][j] is in day i, finish transaction j, the best benefit.
    // The price in day i is prices[i - 1], the transaction j means buy j/2 + j%2 times, and sell j/2 times.
    int[][] dp = new int[m + 1][n + 1];
    dp[0][0] = 0;
    for (int i = 1; i <= m; ++i) {
      for (int j = 1; j <= n && j <= i; ++j) {
        if (j % 2 == 1) {
          // we buy at this time.
          dp[i][j] = dp[i - 1][j - 1] - prices[i - 1];
          if (i - 1 >= j) {
            dp[i][j] = Math.max(dp[i][j], dp[i - 1][j]);
          }
        } else {
          // we sell at this time.
          dp[i][j] = dp[i - 1][j - 1] + prices[i - 1];
          if (i - 1 >= j) {
            dp[i][j] = Math.max(dp[i][j], dp[i - 1][j]);
          }
        }
      }
    }
    int result = 0;
    for (int j = 0; j <= n; ++j) {
      result = Math.max(result, dp[m][j]);
    }
    return result;
  }

  public int maxProfit2(int[] prices) {
    int n = prices.length;
    // dp[0] hold no stock, dp[1] hold one stock.
    int[][] dp = new int[n][2];
    dp[0][0] = 0;
    dp[0][1] = -prices[0];
    for (int i = 1; i < n; ++i) {
      dp[i][0] = Math.max(dp[i - 1][0], dp[i - 1][1] + prices[i]);
      dp[i][1] = Math.max(dp[i - 1][1], dp[i - 1][0] - prices[i]);
    }
    return dp[n - 1][0];
  }
}

File: leetcode/BestTimeToBuyAndSellStockIV.cpp
#include <algorithm>
#include <vector>
using namespace std;
class Solution {
 public:
  int maxProfit(int k, vector<int>& prices) {
    int m = prices.size();
    if (k >= m) {
      return maxProfit2(k, prices);
    }
    int n = std::min(k * 2, m);
    vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));
    for (int i = 1; i <= m; ++i) {
      for (int j = 1; j <= n && j <= i; ++j) {
        if (j % 2 == 1) {
          // we buy at this time.
          dp[i][j] = dp[i - 1][j - 1] - prices[i - 1];
          if (i - 1 >= j) {
            dp[i][j] = max(dp[i][j], dp[i - 1][j]);
          }
        } else {
          // we sell at this time.
          dp[i][j] = dp[i - 1][j - 1] + prices[i - 1];
          if (i - 1 >= j) {
            dp[i][j] = max(dp[i][j], dp[i - 1][j]);
          }
        }
        printf("dp[%d][%d] = %d\n", i, j, dp[i][j]);
      }
    }
    int result = 0;
    for (int j = 0; j <= n; ++j) {
      result = max(result, dp[m][j]);
    }
    return result;
  }
  int maxProfit2(int k, vector<int>& prices) {
    int result = 0;
    for (size_t i = 1; i < prices.size(); ++i) {
      if (prices[i] > prices[i - 1]) {
        result += prices[i] - prices[i - 1];
      }
    }
    return result;
  }
};

int main() {
  Solution solution;
  vector<int> prices = {3,3,5,0,0,3,1,4};
  int result = solution.maxProfit(2, prices);
  printf("result = %d\n", result);
}
